Always set user attributes from shibboleth parameters

soerenbe opened the pull request:
Hello again,
here is a patch for a bug I found during my development.
When the attributes of a shibboleth user changes they will never be reflected in the Django database. This is especially important when the mail address changes. If you want I can make this optional by a special settings parameter like SHIBBOLETH_ALWAYS_UPDATE_USER. But I think there is no requirement for this, because you always want the correct user details in the Django database.


In addition there was also a Django 1.10 deprecated warning:

    /usr/local/share/myproject/shibboleth/backends.py:36: RemovedInDjango110Warning: 'get_all_field_names is an unofficial API that has been deprecated. You may be able to replace it with 'get_fields()'
    shib_user_params = dict([(k, shib_meta[k]) for k in User._meta.get_all_field_names() if k in shib_meta])

I fixed this according to https://docs.djangoproject.com/es/1.9/ref/models/meta/

I also removed some unused libraries at the top.

Best Regards
Sören

bcail commented:
Thanks for this PR. A couple thoughts:
1. Could you add a test for the functionality you're adding (updating user data in the db)?
2. It looks like every time a user logs in, their user object will be retrieved and then saved, whether their Shib attributes have changed or not. Would it be possible to only save the User object if something has changed?

soerenbe commented:
Thanks for the fast feedback.

1.) Ok, I will add this
2.) You are right. I will add this too.

I will work on that next week. I left my shibboleth testbed at work. :-)

soerenbe commented:
I committed the changes to the PR. Some additional notes:

1.) During the test development I found and fixed a bug in the test cases. All other test cases call `auth.authenticate` with the HTTP header. This is wrong because the method requires the valid shibboleth meta data. To fix this I made the method `ShibbolethRemoteUserMiddleware.parse_attributes` a static method.

2.) In the files I touch I made the syntax more pep8 compliant and removed some unused libraries.

bcail commented:
Thanks.
